quadbin/main.py

```diff
@@ -0,0 +1,523 @@
+    bool
+    """
+    header = HEADER
+    mode = (cell >> 59) & 7
+    resolution = (cell >> 52) & 0x1F
+    unused = FOOTER >> (resolution << 2)
+    return (
+        cell >= 0
+        and (cell & header == header)
+        and mode == 1
+        and resolution >= 0
+        and resolution <= 26
+        and (cell & unused == unused)
+    )
+
+
+def cell_to_tile(cell):
+    """Convert a cell into a tile.
+
+    Parameters
+    ----------
+    cell : int
+
+    Returns
+    -------
+    tile: tuple (x, y, z)
+    """
+    # mode = (cell >> 59) & 7
+    # extra = (cell >> 57) & 3
+    z = cell >> 52 & 31
+    q = (cell & FOOTER) << 12
+    x = q
+    y = q >> 1
+
+    x = x & B[0]
+    y = y & B[0]
+
+    x = (x | (x >> S[0])) & B[1]
+    y = (y | (y >> S[0])) & B[1]
+
+    x = (x | (x >> S[1])) & B[2]
+    y = (y | (y >> S[1])) & B[2]
+
+    x = (x | (x >> S[2])) & B[3]
+    y = (y | (y >> S[2])) & B[3]
+
+    x = (x | (x >> S[3])) & B[4]
+    y = (y | (y >> S[3])) & B[4]
+
+    x = (x | (x >> S[4])) & B[5]
+    y = (y | (y >> S[4])) & B[5]
+
+    x = x >> (32 - z)
+    y = y >> (32 - z)
+
+    return (x, y, z)
+
+
+def tile_to_cell(tile):
+    """Convert a tile into a cell.
+
+    Parameters
+    ----------
+    tile : tuple (x, y, z)
+
+    Returns
+    -------
+    int
+    """
+    x, y, z = tile
+
+    x = x << (32 - z)
+    y = y << (32 - z)
+
+    x = (x | (x << S[4])) & B[4]
+    y = (y | (y << S[4])) & B[4]
+
+    x = (x | (x << S[3])) & B[3]
+    y = (y | (y << S[3])) & B[3]
+
+    x = (x | (x << S[2])) & B[2]
+    y = (y | (y << S[2])) & B[2]
+
+    x = (x | (x << S[1])) & B[1]
+    y = (y | (y << S[1])) & B[1]
+
+    x = (x | (x << S[0])) & B[0]
+    y = (y | (y << S[0])) & B[0]
+
+    # -- | (mode << 59) | (mode_dep << 57)
+    return HEADER | (1 << 59) | (z << 52) | ((x | (y << 1)) >> 12) | (FOOTER >> (z * 2))
+
+
+def cell_to_point(cell, geojson=False):
+    """Convert a cell into a geographic point.
+
+    Parameters
+    ----------
+    cell : int
+    geojson : bool, optional
+        Return the point as a GeoJSON geometry, by default False.
+
+    Returns
+    -------
+    str
+        if geojson is True.
+    list
+        if geojson is False (default).
+    """
+    tile = cell_to_tile(cell)
+
+    longitude = tile_to_longitude(tile, 0.5)
+    latitude = tile_to_latitude(tile, 0.5)
+
+    if geojson:
+        return '{{"type": "Point", "coordinates": {0}}}'.format([longitude, latitude])
+    else:
+        return [longitude, latitude]
+
+
+def point_to_cell(longitude, latitude, resolution):
+    """Convert a geographic point into a cell.
+
+    Parameters
+    ----------
+    longitude : float
+        Longitude in decimal degrees.
+    latitude : float
+        Latitude in decimal degrees.
+    resolution : int
+        The resolution of the cell.
+
+    Returns
+    -------
+    int
+
+    Raises
+    ------
+    Exception
+        If the resolution is out of bounds.
+    """
+    if resolution < 0 or resolution > 26:
```

Comment:
We repeat this code (or similar one) several times. Can we consider to encapsulate it into a function `is_valid_resolution` (which I think makes sense) so that we don't need to repeat ourselves in the tests too?

Reply:
It's checked in 3 places, but not exactly the same code, so maybe in this case it's not needed.